database/mongodb/mongodb_test.go

```diff
@@ -0,0 +1,97 @@
+package mongodb
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"testing"
+
+	dt "github.com/golang-migrate/migrate/v4/database/testing"
+	mt "github.com/golang-migrate/migrate/v4/testing"
+	"github.com/mongodb/mongo-go-driver/mongo"
+)
+
+var versions = []mt.Version{
+	{Image: "mongo:4"},
+	{Image: "mongo:3"},
+}
+
+func mongoConnectionString(host string, port uint) string {
+	return fmt.Sprintf("mongodb://%s:%v/testMigration", host, port)
+}
+
+func isReady(i mt.Instance) bool {
+	client, err := mongo.Connect(context.TODO(), mongoConnectionString(i.Host(), i.Port()))
+	if err != nil {
+		return false
+	}
+	defer client.Disconnect(context.TODO())
+	if err = client.Ping(context.TODO(), nil); err != nil {
+		switch err {
+		case io.EOF:
+			return false
+		default:
+			fmt.Println(err)
+		}
+		return false
+	}
+	return true
+}
+
+func Test(t *testing.T) {
+	mt.ParallelTest(t, versions, isReady,
+		func(t *testing.T, i mt.Instance) {
+			p := &Mongo{}
+			addr := mongoConnectionString(i.Host(), i.Port())
+			d, err := p.Open(addr)
+			if err != nil {
+				t.Fatalf("%v", err)
+			}
+			defer d.Close()
+			dt.TestNilVersion(t, d)
+			//TestLockAndUnlock(t, d) driver doesn't support lock on database level
+			dt.TestRun(t, d, bytes.NewReader([]byte(`[{"insert":"hello","documents":[{"wild":"world"}]}]`)))
+			dt.TestSetVersion(t, d)
+			dt.TestDrop(t, d)
+		})
+}
+
+func TestWithAuth(t *testing.T) {
+	mt.ParallelTest(t, versions, isReady,
+		func(t *testing.T, i mt.Instance) {
+			p := &Mongo{}
```

Comment:
split up the test cases e.g. use [table driven tests](https://github.com/golang/go/wiki/TableDrivenTests)

You'll also need to use a different instance of the `Mongo` driver for each test since `Open()` is a pointer receiver method and may potentially modify the driver.

Reply:
Refactored, thank you!